File: pastprograms/0-USACO/b-Silver/2017J-P2.py
fin = open("hps.in", 'r')
fout = open("hps.out", 'w') 
N = int(fin.readline())
past = [(0, 0, 0)]
a, b, c = 0, 0, 0
best = 0
for i in range(N):
    read = fin.readline()[:-1]
    if(read == "P"): a += 1
    elif(read == "H"): b += 1
    else: c += 1
    past.append((a, b, c))
total = [a, b, c]

#P and then H
for cp, ch, cs in past: best = max(best, (total[1] - ch + cp))
#H and then P
for cp, ch, cs in past: best = max(best, (total[0] - cp + ch))
#P and then S
for cp, ch, cs in past: best = max(best, (total[2] - cs + cp))
#S and then P
for cp, ch, cs in past: best = max(best, (total[0] - cp + cs))
#S and then H
for cp, ch, cs in past: best = max(best, (total[1] - ch + cs))
#H and then S
for cp, ch, cs in past: best = max(best, (total[2] - cs + ch))

fout.write(str(best))

File: pastprograms/0-USACO/c-Gold/2017J-P2.cpp
#define pb push_back
#define mt make_tuple
#define mp make_pair
#define is insert
#define ll long long
#define for_loop(i, begin, n) for (ll i = begin; i < n; i ++)
#include <bits/stdc++.h>
using namespace std;

int main () {
	ifstream fin; ofstream fout;
	fin.open("hps.in"); fout.open("hps.out");

	int N, M; 
	fin >> N >> M;
	int H[21][100001], P[21][100001], S[21][100001];
	for_loop (i, 0, 22) {
		H[i][0] = 0;
	}
	for_loop (i, 1, N + 1) {
		char a; fin >> a;
		H[0][i] = H[0][i - 1] + (a == 'H');
		P[0][i] = P[0][i - 1] + (a == 'P');
		S[0][i] = S[0][i - 1] + (a == 'S');
	for_loop (j, 1, M + 1) {
			H[j][i] = max({H[j - 1][i - 1], P[j - 1][i - 1], S[j - 1][i - 1]}) + (a == 'H');
			P[j][i] = max({H[j - 1][i - 1], P[j - 1][i - 1], S[j - 1][i - 1]}) + (a == 'P');
			S[j][i] = max({H[j - 1][i - 1], P[j - 1][i - 1], S[j - 1][i - 1]}) + (a == 'S');
			H[j][i] = max(H[j][i], H[j][i - 1] + (a == 'H'));
			P[j][i] = max(P[j][i], P[j][i - 1] + (a == 'P'));
			S[j][i] = max(S[j][i], S[j][i - 1] + (a == 'S'));
		}
	}
	int best = 0;
	for_loop (i, 0, M + 1) {
		best = max({best, H[i][N], P[i][N], S[i][N]});
	}
	fout << best << endl;
}